src/query/atomWithQuery.ts

```diff
@@ -90,125 +88,101 @@ export function atomWithQuery<
+    const listener = (result: Result) => {
+      if (result.isFetching || (!result.isError && result.data === undefined)) {
+        return
       }
-      const listener = (result: Result) => {
-        if (
-          result.isFetching ||
-          (!result.isError && result.data === undefined)
-        ) {
-          return
-        }
-        if (resolve) {
-          setTimeout(unsubIfNotMounted, 1000)
-          resolve(result)
-          resolve = null
-        } else if (setResult) {
-          setResult(result)
-        } else {
-          throw new Error('setting result without mount')
-        }
+      if (!resolve && !setResult) {
+        throw new Error('setting result without mount')
       }
-      if (options.enabled !== false) {
-        unsubscribe = observer.subscribe(listener)
+      if (resolve) {
+        resolve(result)
+        resolve = null
+      }
+      if (setResult) {
+        setResult(result)
+      }
+      if (result.data !== undefined) {
+        previousDataCache.set(queryClient, result.data)
       }
-      resultAtom.onMount = (update) => {
-        setResult = update
-        if (options.enabled !== false && !unsubscribe) {
+    }
+    let unsubscribe: (() => void) | null = null
+    let timer: Timeout | undefined
+    const startQuery = () => {
+      if (!setResult && unsubscribe) {
+        clearTimeout(timer)
+        unsubscribe()
+        unsubscribe = null
+      }
+      if (options.enabled !== false) {
+        if (setResult) {
+          observer.refetch({ cancelRefetch: true }).then(setResult)
+        } else {
           unsubscribe = observer.subscribe(listener)
-          listener(observer.getCurrentResult())
-        }
-        return () => {
-          setResult = null
-          unsubscribe?.()
         }
       }
-      return { options, resultAtom, unsubIfNotMounted }
-    },
-    (get, set, action) => {
-      const observer = get(observerAtom)
-      const { options, resultAtom, unsubIfNotMounted } = get(queryDataAtom)
-      if (options.enabled === false) {
-        return
+      if (!setResult) {
+        // not mounted yet
+        timer = setTimeout(() => {
+          if (unsubscribe) {
+            unsubscribe()
+            unsubscribe = null
+          }
+        }, 1000)
       }
-      switch (action.type) {
-        case 'refetch': {
-          set(resultAtom, new Promise<never>(() => {})) // infinite pending
-          unsubIfNotMounted()
-          return observer.refetch({ cancelRefetch: true }).then((result) => {
-            set(resultAtom, result)
-          })
+    }
+    startQuery()
+    resultAtom.onMount = (update) => {
+      setResult = update
+      if (unsubscribe) {
+        clearTimeout(timer)
+      } else {
+        startQuery()
+      }
+      return () => {
+        setResult = null
+        if (unsubscribe) {
+          unsubscribe()
+          // FIXME why does this fail?
+          // unsubscribe = null
```

Comment:
I wasn't be able to solve this. Maybe I'm missing something. Happy if someone tries it.

Reply:
Ah, maybe it's Strict Effect's double invocation.